### PR TITLE
Stud 6 all courses page front end

### PR DESCRIPTION
**Pull request by el-sharif-mohamed22**
### All Courses Page
All courses are displayed after grouping them in terms.
The search bar works alongside the filter by status.
Add course functionality with reactive form.
Delte course popmenu.

**This page works with dummy data from a JSON file and will be modified after handling it in the back end.**

### Components created:
* All-course component.
* Course card component.
* Add course component.

### Services created:
* All course services, for loading the courses.
* Add course service, for sending a request to add a course.
* Delete course service, for deleting a course.

### Modifications:
* Course model: add a new attribute "term".
* Navigation bar: handled the smoothness of the animation.

### CODE REVIEW

File: Student-Unity-frontend/src/app/components/navigation-bar/navigation-bar.component.ts [hunk 1]
```diff
@@ -18,8 +18,16 @@ export class NavigationBarComponent implements OnInit {
     this.getSignedInUser();
   }
 
-  navigateTo(child: string){
+  navigateTo(event: any, child: string){
     this.router.navigateByUrl("home/" + child);
+    let el = event.target;
```
Comment (Ahmedelsa3eed) [inline Student-Unity-frontend/src/app/components/navigation-bar/navigation-bar.component.ts:23]: what's going here

File: Student-Unity-frontend/src/app/components/all-courses/all-courses.component.ts [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+          console.log(this.privilege);
+        }
+    },
+      err => {
+        console.log(err);
+      }
+      );
+  }
+
+  filterCourses(): void {
```
Comment (Ahmedelsa3eed) [inline Student-Unity-frontend/src/app/components/all-courses/all-courses.component.ts:106]: i'm not sure about filtering the course on the client side

File: Student-Unity-frontend/src/app/models/Course.ts [hunk 3]
```diff
@@ -5,4 +5,5 @@ export interface Course {
   telegramLink: string;
   status: boolean;
   notificationsToken: string;
+  term: number;
```
Comment (Ahmedelsa3eed) [inline Student-Unity-frontend/src/app/models/Course.ts:8]: this column is not mentioned in the desing

**Review by Ahmedelsa3eed — COMMENTED** [review]
i think we could combine the three services into one service as all of them relates to courses anyway 
**keep the hard work man**

**Comment by Ahmedelsa3eed** [comment]
@el-sharif-mohamed22 
- please resolve these conflicts
- you may need to change ``courses: Course[] = []`` to courses?: Observalbe<Course[]>``, so that when a course attribute changes, its effect will be shown directly in the template.
- you can look at ``all-tasks`` component for more details.

File: Student-Unity-frontend/src/app/components/navigation-bar/navigation-bar.component.ts (hunk 1 above)
Comment (el-sharif-mohamed22) [reply Student-Unity-frontend/src/app/components/navigation-bar/navigation-bar.component.ts]: Fixed this issue using router active for coloring the button blue when the URL changes.

File: Student-Unity-frontend/src/app/components/all-courses/all-courses.component.ts (hunk 2 above)
Comment (el-sharif-mohamed22) [reply Student-Unity-frontend/src/app/components/all-courses/all-courses.component.ts:106]: This filtering function is only useful for separating the courses in the view, it doesn't have other uses. So one call from the back-end to get all the courses and then display them in a different view.

**Comment by el-sharif-mohamed22** [comment]
Using `courses: Course[] = []` creates an empty array of courses, So when the request fails this object is still empty which in return displays nothing. but using `courses?: Course[]` can lead that this object being undefined and then being accessed in different methods which leads to errors.

**Comment by el-sharif-mohamed22** [comment]
> i think we could combine the three services into one service as all of them relates to courses anyway **keep the hard work man**

Combined the 3 services into one service

**Review by Ahmedelsa3eed — APPROVED** [review]

**Review by AhmedElemary57 — APPROVED** [review]

**Pull request merged**